core/services/wifi/WifiManager.py

```diff
@@ -113,78 +84,125 @@ def __dict_from_list(data: bytes) -> Dict[str, Any]:
-            scan_data = self.wpa.send_command_scan()
-            if b"BUSY" in scan_data:
-                raise BusyError
-            data = self.wpa.send_command_scan_results()
-            networks_list = WifiManager.__dict_from_table(data)
-            return [ScannedWifiNetwork(**network) for network in networks_list]
-        except BusyError as error:
-            raise BusyError("WPA service is busy. Try to space network scans by at least 4 seconds.") from error
-        except Exception as error:
-            raise FetchError("Failed to fetch wifi list.") from error
 
-    def get_saved_wifi_network(self) -> List[SavedWifiNetwork]:
+        async def perform_new_scan() -> None:
+            try:
+                self._scan_task = asyncio.create_task(self.wpa.send_command_scan(timeout=15))
+                await self._scan_task
+                data = await self.wpa.send_command_scan_results()
+                networks_list = WifiManager.__dict_from_table(data)
+                self._updated_scan_results = [ScannedWifiNetwork(**network) for network in networks_list]
+            except Exception as error:
+                if self._scan_task is not None:
+                    self._scan_task.cancel()
+                self._updated_scan_results = None
+                raise FetchError("Failed to fetch wifi list.") from error
+
+        # Performs a new scan only if it's the first one or the last one is already done
+        # In case there's one running already, wait for it to finish and use its result
+        if self._scan_task is None:
+            await perform_new_scan()
+        elif self._scan_task.done():
+            await perform_new_scan()
+        else:
+            awaited_scan_task_name = self._scan_task.get_name()
+            while self._scan_task.get_name() == awaited_scan_task_name and not self._scan_task.done():
+                logger.info(f"Waiting for {awaited_scan_task_name} results.")
+                await asyncio.sleep(0.5)
+
+        if self._updated_scan_results is None:
+            raise FetchError("Failed to fetch wifi list.")
+
+        return self._updated_scan_results
+
+    async def get_saved_wifi_network(self) -> List[SavedWifiNetwork]:
         """Get a list of saved wifi networks"""
         try:
-            data = self.wpa.send_command_list_networks()
+            data = await self.wpa.send_command_list_networks()
             networks_list = WifiManager.__dict_from_table(data)
             return [SavedWifiNetwork(**network) for network in networks_list]
         except Exception as error:
             raise FetchError("Failed to fetch saved networks list.") from error
 
-    def set_wifi_password(self, credentials: WifiCredentials) -> Any:
+    async def add_network(self, credentials: WifiCredentials, hidden: bool = False) -> int:
         """Set network ssid and password
 
         Arguments:
             credentials {[WifiCredentials]} -- object containing ssid and password of the network
```

Comment:
missing the hidden argument in the docstring

Reply:
Fixed.